ssc/libssc.c

```diff
@@ -610,84 +873,177 @@ static PetscErrorCode PCPatchCreateCellPatchBCs(PC pc,
             PetscHashIHasKey(globalBcs, globalDof, flg);
             if (flg) {
                 PetscHashIAdd(localBcs, localDof, 0);
+                PetscHashIAdd(multLocalBcs, localDof, 0);
             }
         }
-        ierr = PetscSectionGetDof(facetCounts, v, &dof); CHKERRQ(ierr);
-        ierr = PetscSectionGetOffset(facetCounts, v, &off); CHKERRQ(ierr);
-        for ( PetscInt i = off; i < off + dof; i++ ) {
-            const PetscInt f = facetsArray[i];
-            ierr = DMPlexGetTransitiveClosure(dm, f, PETSC_TRUE, &closureSize, &closure); CHKERRQ(ierr);
-            for ( PetscInt ci = 0; ci < closureSize; ci++ ) {
-                PetscInt ldof, loff;
-                const PetscInt p = closure[2*ci];
-                ierr = PetscSectionGetDof(dofSection, p, &ldof); CHKERRQ(ierr);
-                ierr = PetscSectionGetOffset(dofSection, p, &loff); CHKERRQ(ierr);
-                if ( ldof > 0 ) {
-                    for ( PetscInt j = loff; j < ldof + loff; j++ ) {
-                        PetscInt localDof;
-                        PetscHashIMap(patchDofs, j, localDof);
-                        if ( localDof == -1 ) {
-                            SETERRQ(PETSC_COMM_SELF, PETSC_ERR_ARG_OUTOFRANGE,
-                                    "Didn't find facet dof in patch dof\n");
-                        }
-                        PetscHashIAdd(localBcs, localDof, 0);
-                    }
+
+        /* If we're doing multiplicative, make the BC data structures now
+           corresponding solely to actual globally imposed Dirichlet BCs */
+        if (patch->multiplicative) {
+            PetscHashISize(multLocalBcs, numBcs);
+            ierr = PetscSectionSetDof(multBcCounts, v, numBcs); CHKERRQ(ierr);
+            ierr = PetscMalloc1(numBcs, &multBcsArray); CHKERRQ(ierr);
+            ierr = PetscHashIGetKeys(multLocalBcs, &multBcIndex, multBcsArray); CHKERRQ(ierr);
+            ierr = PetscSortInt(numBcs, multBcsArray); CHKERRQ(ierr);
+            ierr = ISCreateGeneral(PETSC_COMM_SELF, numBcs, multBcsArray, PETSC_OWN_POINTER, &(patch->multBcs[v - vStart])); CHKERRQ(ierr);
```

Comment:
I think we can just use the `localBcs` hash table for this?  Maybe check (this could save minimal memory).

Reply:
Fixed.